### DISCUSSION TITLE
Seperation between dataclass args and sqlalchemy args in definition

### DISCUSSION THREAD

**Discussion by gam-phon** [sqlalchemy/sqlalchemy#9795] [Ideas]
Hello,

I like the approach of separating between dataclass args and sqlalchemy args in definition : for example ```default``` and ```insert_default``` has a deep meaning of applying to sqlalchemy or dataclass. Also with the time you have to understand deeply witch arg. affect which one either dataclass or sqlachemy or maybe both. Also the Ordering of dataclass default is an issue where as in sqlalchemy is not. 

**Explicit is better than implicit**. I like things to be more explicit this is why i prefer ```back_populates``` than using ```backref```

I am discussing this to get an overview of the best approach to achieve sqlalchemy and dataclass in the same model. if you check #9410 you will see that there are some issues with the current approach, mixing them together.
I got inspired by the approach below by this issue #9785 

Example:
```python
id: Mapped[str] = mapped_column(String, nullable=False, repr=True)
name: Mapped[str] = mapped_column(String, nullable=False, insert_default="", default="You", repr=True)
```
if name ```name``` is above ```id``` it will not work and you will get an error. below will not work
```python
name: Mapped[str] = mapped_column(String, nullable=False, insert_default="", default="You", repr=True)
id: Mapped[str] = mapped_column(String, nullable=False, repr=True)
```
## First approach (working now)
Maybe Could be like this to make sure and clear things are seperated also you could include and exclude some fileds by this way
```python
from dataclasses import field
from typing import TYPE_CHECKING

if not TYPE_CHECKING:
    # Here order of fields are not important
    name = mapped_column(String, nullable=False, default="") # also no need for insert_default
    id = mapped_column(String, nullable=False, default="")
    secret = mapped_column(String, nullable=False, default="") # you could use InitVar to achieve this as well but it will not be saved in DB
else:
    # Here order of fields is important, regular dataclass fields
    id: Mapped[str] = field(default="You", repr=True)
    name: Mapped[str] = field(default="You", repr=True)
    # secret will be excluded. since asdict does not support excluding things, for now, this is a wrokaround but you will not get autocompletion for it, this is the trade off ^_^ 
```
## Second approach (needs implementation)

Or maybe creating a specific field in mapped_column that will fill dataclass filed example and making order is not important with this things will become more clear which args belong to which
```python
name: Mapped[str] = mapped_column(String, nullable=False, default="", dataclass_opts={"default"="You", "repr"=True})
id: Mapped[str] = mapped_column(String, nullable=False, dataclass_opts={"repr"=True})
secret: Mapped[str] = mapped_column(String, nullable=False, default=""dataclass_opts=None) # will not be included in dataclass
```
I think this approach will make things more slower and will cause compatibility issues

Edit: Approach 1 is not working actually ^_^

**Comment by CaselIT** [comment]
Hi,

I don't think what you propose is viable for several reasons:
- if you add "MappedAsDataclass" it means that you want dataclasses at runtime, so you can't use `if TYPE_CHECKING` otherwise the generated classes will not be dataclasses. Your first approach that you call "working now" does not really work, since all the default are not you bug empty string. Also any option you pass, such as `init=False` would not be respected at runtime.
- how dataclass fields are declared is defined by https://peps.python.org/pep-0681/, so we cannot change it by adding `dataclass_opts` or similar of you loose all typing support
- I think that an user chooses to use dataclasses then they need to be aware of how they behave and what their limitation are. If that's not what they want than they can avoid using dataclasses

**Reply by gam-phon** [reply]
I got it now

- ```TYPE_CHECKING``` is True used in 3rd party static type checkers. 
- ```not TYPE_CHECKING``` is False used in runtime

so the first approach at runtime will run ```not TYPE_CHECKING``` and evaluate dataclass as will.

second approach not compatible with dataclass definition

Thanks @CaselIT

**Reply by CaselIT** [reply]
> second approach not compatible with dataclass definition

it's not compatible with how type checker interpret a class as being a dataclass, so they would not know that `dataclass_opts={"init":False}` means that the field is not in the init.
SQLAlchemy could support something like this without much issue, but it would be of limited value, since all type checking would be broken

**Reply by zzzeek** [reply]
you can't use an arbitrary field order at runtime, Python dataclasses are a runtime thing.  We pass the fields to dataclasses in the order they are received.

I'm not really seeing what the problem is with things right now, "back_populates" is not much of  a problem only `asdict()/astuple()` are.  for ordering etc., we are building up an `__init__()` method, that's the whole thing we're doing here, so that's why order is and should be significant.   if you set `init=False` on the class, ordering can be anything.
